### PR TITLE
Add full delta support for sum aggregator

### LINKED ISSUE #3160: Add delta support for sum aggregator for SumObserver and UpDownSubObserver
**Is your feature request related to a problem? Please describe.**
The `opentelemetry-java-instrumentation` runtime metrics use UpDownSumObserver instruments to observe jvm memory usage. Currently, there's no way to configure the agent to export delta sums from these instruments because its not supported in the SDK.

**Describe the solution you'd like**
The current sum aggregator only partially supports delta aggregation temporality. This is reflected in the factory method which accepts `boolean alwaysCumulative` instead of `AggregationTemporality temporality`: 
```
AggregatorFactory sum(boolean alwaysCumulative) 
``` 
This behavior reflects the fact that sum observer and up down sum observer are assumed to be observing cumulative sum metrics. Converting observed cumulative sums to deltas is possible, but requires maintaining state and taking a diff of `currentAccumulation - previousAccumulation`. 

I'd like to see full delta support added to the sum aggregator factory, and this be reflected in a new constructor signature:
```
static AggregatorFactory sum(AggregationTemporality temporality)
```

**Describe alternatives you've considered**
None. 

**Additional context**
None.

### PR DESCRIPTION
**Pull request by jack-berg**
Closes #3160.

### CODE REVIEW

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/Aggregator.java [hunk 1]
```diff
@@ -57,9 +57,11 @@ default T accumulateDouble(double value) {
   /**
    * Returns the result of the merge of the given accumulations.
    *
+   * @param previousAccumulation the previously captured accumulation
+   * @param accumulation the newly captured accumulation
    * @return the result of the merge of the given accumulations.
    */
-  T merge(T a1, T a2);
+  T merge(T previousAccumulation, T accumulation);
```
Comment (jkwatson) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/Aggregator.java:64]: 👍🏼

**Review by jkwatson — APPROVED** [review]
Thanks! This looks good to me.

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractSumAggregator.java [hunk 2]
```diff
@@ -14,30 +14,92 @@
...
+   * Resolve the aggregator merge strategy. The merge strategy is SUM in all cases except where
+   * temporality is {@link AggregationTemporality#DELTA} and instrument type is {@link
+   * InstrumentType#SUM_OBSERVER} or {@link InstrumentType#UP_DOWN_SUM_OBSERVER}. In these special
+   * cases, the observed values are cumulative sums so we must take a diff to compute the delta sum.
+   *
+   * @param instrumentType the instrument type
+   * @param temporality the temporality
+   * @return the merge strategy
+   */
+  static MergeStrategy resolveMergeStrategy(
```
Comment (anuraaga) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractSumAggregator.java:67]:
```suggestion
  // Visible for testing
  static MergeStrategy resolveMergeStrategy(
```

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractSumAggregator.java [hunk 3]
```diff
@@ -14,30 +14,92 @@
...
+  @Override
+  public final T merge(T a1, T a2) {
+    switch (mergeStrategy) {
+      case SUM:
+        return mergeSum(a1, a2);
+      case DIFF:
+        return mergeDiff(a1, a2);
     }
-    this.temporality = temp;
+    throw new AssertionError("Unsupported merge strategy: " + mergeStrategy.name());
```
Comment (anuraaga) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractSumAggregator.java:86]:
```suggestion
    throw new IllegalStateException("Unsupported merge strategy: " + mergeStrategy.name());
```

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractSumAggregator.java [hunk 4]
```diff
@@ -14,30 +14,92 @@
...
+            || instrumentType == InstrumentType.UP_DOWN_SUM_OBSERVER)
+        && temporality == AggregationTemporality.DELTA) {
+      return MergeStrategy.DIFF;
+    } else {
+      return MergeStrategy.SUM;
+    }
+  }
+
+  @Override
+  public final T merge(T a1, T a2) {
```
Comment (anuraaga) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/AbstractSumAggregator.java:79]: Let's apply the variable name fix here too, and all similar methods in this PR

File: sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/DoubleSumAggregator.java [hunk 5]
```diff
@@ -35,10 +35,15 @@ public Double accumulateDouble(double value) {
   }
 
   @Override
-  public final Double merge(Double a1, Double a2) {
+  final Double mergeSum(Double a1, Double a2) {
```
Comment (anuraaga) [inline sdk/metrics/src/main/java/io/opentelemetry/sdk/metrics/aggregator/DoubleSumAggregator.java:38]: Don't need `final` on methods of `final` class, throughout PR.

**Review by anuraaga — COMMENTED** [review]
Just some small points, thanks

**Comment by jack-berg** [comment]
Thanks for the feedback @anuraaga! All set on those changes.

**Review by anuraaga — APPROVED** [review]
Thanks!

**Pull request merged**